### PR TITLE
i2c for esp32-c3

### PR DESCRIPTION
**Pull request by zdima**
Added support for I2C with 7-bit addressing for ESP32C3.
Tested with bh1750 (reads) and SSD1306 (writes).

### CODE REVIEW

File: src/machine/i2c.go [hunk 1]
```diff
@@ -1,5 +1,4 @@
-//go:build atmega || nrf || sam || stm32 || fe310 || k210 || rp2040
-// +build atmega nrf sam stm32 fe310 k210 rp2040
+// +build atmega nrf sam stm32 fe310 k210 rp2040 esp32c3
```
Comment (sago35) [inline src/machine/i2c.go:2]: Please do not delete definitions from go 1.17 or later.

```suggestion
//go:build atmega || nrf || sam || stm32 || fe310 || k210 || rp2040 || esp32c3
// +build atmega nrf sam stm32 fe310 k210 rp2040 esp32c3
```

**Comment by sago35** [comment]
@zdima 
I tried with the ADT7410 example.
However, m5stamp-c3 did not produce the expected waveform.
feather-m4-can (atsamd51) is working correctly.
Can you think of any possible causes?

m5stamp-c3
image

feather-m4-can
image

```go
package main

import (
	"fmt"
	"machine"
	"time"

	"tinygo.org/x/drivers/adt7410"
)

var (
	i2c    = machine.I2C0
	sensor = adt7410.New(i2c)
)

func main() {
	i2c.Configure(machine.I2CConfig{Frequency: machine.TWI_FREQ_400KHZ, SCL: machine.IO6, SDA: machine.IO7})
	//i2c.Configure(machine.I2CConfig{Frequency: machine.TWI_FREQ_400KHZ})
	sensor.Configure()

	for {
		temp := sensor.ReadTempC()
		fmt.Printf("temperature: %f\r\n", temp)
		time.Sleep(time.Second)
	}

}
```

**Comment by sago35** [comment]
I rewrote it as follows, and it works a little better.
But it still doesn't work well.

m5stamp-c3
image

```go
func main() {

	i2c := machine.I2C0
	i2c.Configure(machine.I2CConfig{Frequency: machine.TWI_FREQ_400KHZ, SCL: machine.IO6, SDA: machine.IO7})
	//i2c.Configure(machine.I2CConfig{Frequency: machine.TWI_FREQ_400KHZ})
	sensor := adt7410.New(i2c)
	sensor.Configure()

	for {
		temp := sensor.ReadTempC()
		fmt.Printf("temperature: %f\r\n", temp)
		time.Sleep(time.Second)
	}

}
```

**Comment by sago35** [comment]
I have confirmed that the SSD1306 does work.

File: src/machine/i2c.go (hunk 1 above)
Comment (zdima) [reply src/machine/i2c.go:2]: my bad. I have fix it.

**Comment by zdima** [comment]
@sago35 

> @zdima I tried with the ADT7410 example. However, m5stamp-c3 did not produce the expected waveform. feather-m4-can (atsamd51) is working correctly. Can you think of any possible causes?

From the wave form it look like the signal capture is partial. The signal with address is not captured and only data portion shown.

> I rewrote it as follows, and it works a little better.

I think this is because sensor is initialized with i2c interface before i2c get initialized. This is only thing that may provide unexpected result. In case you need to keep sensor as a global variable, you could add function that return configured i2c interface and use it to initialize sensor while program is loading.

**Comment by sago35** [comment]
@zdima 
I did some more research.

---

The following source code cannot be loaded by m5stamp-c3.
It looks like there is no address transmission on read.


```go
		i2c.Tx(0x48, []byte{0x00}, buf[:2])
```

m5stamp-c3
image

feather-m4-can
image


---

In the case of the following source code, m5stamp-c3 can also read the data of ADT7410.


```go
		i2c.Tx(0x48, []byte{0x00}, []byte{})
		i2c.Tx(0x48, []byte{}, buf[:2])
```

m5stamp-c3
image

feather-m4-can
image

**Comment by sago35** [comment]
@zdima can you please rebase this branch against dev?

**Comment by zdima** [comment]
@sago35 I don't have I2C device with registers. Can you try new code and let me know if this fix the issue?

**Comment by sago35** [comment]
@zdima 

The waveform is now good enough to communicate.

image

However, there are some waveform changes that are too fast, perhaps due to timing issues with register operations.

image

image


In addition, this process reads 2 bytes, but the second byte is always 0xFF.
I don't understand what the following source code is doing, but it seems to me that only one byte is being processed.

https://github.com/zdima/tinygo/blob/2bfdd894e76138efca8b3738cc22a9bd1a85d089/src/machine/machine_esp32c3.go#L409-L411

**Comment by zdima** [comment]
> However, there are some waveform changes that are too fast,

The registry operations are all complete before it start sending the entire sequence. I don't think I can do anything with this.

> I don't understand what the following source code is doing

The reason for this code is to set for last byte NACK. 

> this process reads 2 bytes, but the second byte is always 0xFF.

Yes, looks like because of this twist with the last byte, the last byte was not getting read from the controller.
Please try the last changes.

**Comment by sago35** [comment]
ADT7410 is now working.
Thank you for the quick response.

File: src/machine/machine_esp32c3.go [hunk 2]
```diff
@@ -152,3 +156,324 @@ func (uart *UART) WriteByte(b byte) error {
...
+	}
+	return nil
+}
+
+// Tx does a single I2C transaction at the specified address.
+// It clocks out the given address, writes the bytes in w, reads back len(r)
+// bytes and stores them in r, and generates a stop condition on the bus.
+func (i2c *I2C) Tx(addr uint16, w, r []byte) (err error) {
+	// timeout in microseconds.
+	const timeout = 100 * 1000 // 40ms is a reasonable time for a real-time system.
```
Comment (sago35) [inline src/machine/machine_esp32c3.go:466]: Probably 40 is correct here.

```suggestion
	const timeout = 40 // 40ms is a reasonable time for a real-time system.
```

**Comment by sago35** [comment]
@zdima Now, it looks like it is failing to read a single byte.
Here, we want to read only one byte from address==0xA1, but it keeps reading until timeout.

https://github.com/tinygo-org/drivers/blob/a15f2167cc7e46864d82e08904a3ddaa3e610d5e/bme280/bme280.go#L65-L69

image

**Comment by osechet** [comment]
Hi @zdima!
is there any news about this PR? I just tested it with a [Lolin C3 Mini v1](https://www.wemos.cc/en/latest/c3/c3_mini_1_0_0.html) and an SSD1306 OLED display and it works perfectly. The test program is visible here if it can help: https://gist.github.com/osechet/d6d3e2d3bc8ad33435705c3a94dc85af.

The program is flashed using the _esp32c3_ target: `tinygo flash -port COM3 -target esp32c3 .\ssd1306`
tinygo version: 0.27.0 windows/amd64
go version: 1.19.4

**Comment by deadprogram** [comment]
I am also very interested in finding out the status of this PR and helping out however needed. :smile_cat:

**Comment by deadprogram** [comment]
Just submitted PR https://github.com/tinygo-org/tinygo/pull/4011 

What do we think is needed if anything to get this PR completed?

**Comment by deadprogram** [comment]
I did some testing last night.

The good news is that with a couple small changes this branch worked to read/write data.
The not as good news is that as @sago35 reported the PR does not work correctly when writing a single byte. I will look into that this evening to see if I can resolve.

**Comment by aykevl** [comment]
Looks like the code here has been merged already in #4014 .

**Pull request closed without merging**